# sumQC

sumQC is a bioinformatics pipeline for cleaning 2nd and 3rd generation sequencing data. In addition, the program also gives to the user a summary table of the sequence data quality before and after the cleaning in the tab delimited text format, ready to be used for publication. The pipeline is written using the bioinformatics workflow manager [Nextflow](https://www.nextflow.io/), a powerful and flexible workflow language that allows multiple tasks at different processing levels to be run parallely.
 
 
sumQC contains two separate workflows; one for processing short- read data generated by 2nd generation sequencing technologies, and the other for processing and long read sequences generated by 3rd generation sequencers, utilising different sets of programs as shown below: 
  
  
|Workflow|QC|Cleaning|
|:--:|:--:|:--:|
|Short read|[FastQC](https://github.com/s-andrews/FastQC)|[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)|
|Long read|[LongQC](https://github.com/yfukasawa/LongQC)|[FiltLong](https://github.com/rrwick/Filtlong)|


In this current version, sumQC cannot process short- and long-read sequence data at the same time, and the user has to specify whether they want to process short or long reads. 

***

# Pipeline summary 

1) Quality check before cleaning by. `fastqc` and `multiqc` (for short reads) or `LongQC` (for long reads)
2) Extracting data quality statistics before the cleaning.
3) Cleaning the input fastq files by. `Trimmomatic` (for short reads) or `FiltLong`(for long reads)
4) Quality check after the before cleaning, again by. `fastqc` and `multiqc` (for short reads) or `LongQC` (long reads)
5) Extracting data quality statistic after the before cleaning.
6) Making a table of data quality summary statistics. Summarise quality statistic into one table.

***

# Requirements

Most of the dependencies required by sumQC are already included in this repository. However, user would still need to install the following programs to use sumQC using the commands shown below:

1. Java 8 or later
2. [Nextflow](https://www.nextflow.io/) (>=21.04.0)
3. [Multiqc](https://github.com/ewels/MultiQC.git)
```
# Install through pip
pip install multiqc

# Install through conda
conda install -c bioconda multiqc
```
4. Python library dependency for LongQC (please check [LongQC github page](https://github.com/yfukasawa/LongQC) for more details)
```
# Install through pip
pip install numpy scipy matplotlib scikit-learn pandas jinja2 h5py pysam edlib python-edlib

# Install through conda
conda install h5py
conda install -c bioconda pysam
conda install -c bioconda edlib
conda install -c bioconda python-edlib

```

***

# Installation 
sumQC can only be installed and used on UNIX/Linux. To install sumQC, use the following command shown to download this repository to your computer from GitHub
```
git clone https://github.com/thavinb/sumqc.git
```
The program comes with the wrapper name 'sumqc'. To test run the program, simply go to the downloaded directory and run the following command
```
./sumqc -h 
```
You should see a help page of the program. For convenience, consider putting the program in your PATH environment variable.

***





# Basic usage 

At the moment, the user can only process either short or long reads at a time. 

For paired-end short-read data, use the following command:
```
./sumqc SR -t PE -i <path/to/shortread.fastq.gz> -o <path/for/output/directory>
```
For single-end short-read data, use the following command:
```
./sumqc SR -t SE -i <path/to/shortread.fastq.gz> -o <path/for/output/directory>
```
For long-read data, the user must specify the sequencing platform through the command option “-p”, and use the following command:
```
./sumqc LR -p <pb-rs2|pb-sequel|ont-ligation|ont-rapid|ont-1dsq> -i <path/to/shortread.fastq.gz> -o <path/for/output/directory>
```
To specify the fastq input files through the command option `-i`, use a glob pattern. The default values are `*[12].fastq.gz` for the paired-end short-read mode, and `*.fastq.gz` for the single-end short-read or long-read modes. The output directory can be specified using option `-o`. User can also specify how they wish Trimmomatic or FiltLong to clean their data through option `-q`; See help page for more details.
   
Followings are all available options: 

```
-h      Print this help page

-i      Input fastq files, specified by using a glob pattern
        (e.g. *[12].fastq.gz)

-o      Output directory

-m      For paired-end reads only. Merge unpaired 
        forward and reverse reads into one file 
        after the cleaning process.

-p      Sequencing platform for long-read sequence processing.
        (https://github.com/yfukasawa/LongQC)

        Platforms           Options
        ---------           -------
        RS-II               pb-rs2
        Sequel              pb-sequel
        ONT(1D ligation)    ont-ligation
        ONT(rapid)          ont-rapid

-q      Cleaning options 

        default value for short-read mode is:
        “SLIDINGWINDOW:4:30 MINLEN:70”

        default value for long-read mode is:
        “min_length 10000 --keep_percent 90”
        
        Please refer to the program webpages for the syntax and the full list of
        options.

-r      Resume from cache in the work directory.
        This is the same as:

        nextflow run main.nf -resume.

        (https://www.nextflow.io/docs/latest/cli.html?highlight=resume#run)
```

***

# Output 

In addition to clean fastq files, sumQC also returns a summary statistics table of the data quality, both before and after cleaning. 
An example of the table is shown below: 

|RAW|||||||||CLEANED|||||||||||||||||
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|FWD|||||RVS||||PAIRED_FWD||||PAIRED_RVS||||UNPAIRED_FWD||||UNPAIRED_RVS|||||
|FileName|TotalSeq|Length|GC|AvgQScore (min,max)|TotalSeq|Length|GC|AvgQScore (min,max)|TotalSeq|Length|GC|AvgQScore (min,max)|TotalSeq|Length|GC|AvgQScore (min,max)|TotalSeq|Length|GC|AvgQScore (min,max)|TotalSeq|Length|GC|AvgQScore (min,max)|Drop|
|test_1-1T|10000|150|65.000|33.828 (27,37)|10000|150|66.000|31.439 (26,37)|2636 (26.36)|70-150|62.000|35.872 (34,37)|2636 (26.36)|70-150|61.000|35.483 (33,37)|3796 (37.96)|70-150|65.000|35.606 (34,37)|819 (8.19)|70-150|61.000|35.258 (33,37)|10113 (50.565)|
|test_3-1T|10000|150|66.000|33.917 (27,37)|10000|150|66.000|31.542 (26,37)|2900 (29)|70-150|63.000|35.941 (34,37)|2900 (29)|70-150|62.000|35.490 (33,37)|3748 (37.48)|70-150|66.000|35.676 (34,37)|784 (7.84)|70-150|62.000|35.310 (33,37)|9668 (48.34)|

***

# Acknowledgements 

An extensive list of references for the tools used in the pipeline can be found in the CITATIONS.md file.


